### PR TITLE
Video block: Gracefully auto-transform into Embed if embeddable resource provided

### LINKED ISSUE #4922: Add failure information to the video block
This came up in usability testing feedback here: https://design.oit.ncsu.edu/2018/02/02/gutenberg-testing/

> Users don’t initially make the connection that “where my video is hosted” has anything to do with “what block I should choose.” Some had to be told to try a different block after multiple attempts to get the Video block to work.

> Blocks should tell you when they break. Rather than displaying an empty video player in the editor (that could be interpreted as a placeholder for the real thing!), URL validation could have detected that the URL provided by the user wouldn’t work and generated a message saying something went wrong.

Can we add better failure information?

### PR DESCRIPTION
**Pull request by mcsf**
Partly fixes #4922. This is an experimental take on the parent issue which eschews the display of failure information. More than anything else, it needs feedback on whether it's a feature worth having, and—if so—it needs design feedback. Moreover, adding this automated correction behavior doesn't exclude displaying failure information if we can't handle the video automatically.

## Description

When adding a new Video block, if the URL provided doesn't correspond to a usable video file (or if its loading otherwise fails) _and_ the URL corresponds to an embeddable resource, then automatically transform the block into a proper Embed block.

gutenberg-video-to-embed

## How Has This Been Tested?

Try adding a Video block with the following kinds of values:
- A valid video URL (e.g. `https://upload.wikimedia.org/wikipedia/commons/9/96/Curiosity%27s_Seven_Minutes_of_Terror.ogv`)
- A valid video URL while the network is down
- An invalid video URL (e.g. a mis-copied link like `https://upload.wikimedia.org/wikipedia/commons/9/96/Curiosity%27s_Seven_Minutes_of_Terror.og`)
- A valid embeddable URL (e.g. `https://youtube.com/watch?v=lNHhmz-smlE`)
- An invalid embeddable URL (e.g. `https://youtube.com/watch?v=lNHhmz-sml`)

Verify that the block is only converted to Embed when provided a valid embeddable URL.

## Types of changes
Tentative UX enhancement

## A technical aside

This pull request refactors the Video block's `edit` method into a functional component. For this, it extended `withState` in order to accommodate its use case, allowing prop-dependent state initializations. If this is a desirable addition, I can spawn a new PR for it.

Because of the refactor, the [Commits](https://github.com/WordPress/gutenberg/pull/4990/commits) view may provide diffs that are easier to review.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style.
- [ ] My code has proper inline documentation.

### CODE REVIEW

File: blocks/library/video/index.js [hunk 1]
```diff
@@ -57,116 +66,172 @@ export const settings = {
...
+				return;
+			}
+
+			const apiURL = addQueryArgs( wpApiSettings.root + 'oembed/1.0/proxy', {
+				url: src,
+				_wpnonce: wpApiSettings.nonce,
+			} );
+
+			console.log( 'tryEmbed; querying' );
+			window.fetch( apiURL, { credentials: 'include' } ).then(
```
Comment (youknowriad) [inline blocks/library/video/index.js]: Should we transform this to class component to avoid calling `setState` on unmounted components?

**Comment by youknowriad** [comment]
Waiting for an UX/design review to do a more in-depth code review.

File: blocks/library/video/index.js (hunk 1 above)
Comment (mcsf) [reply blocks/library/video/index.js]: Either that, or we source a bound `fetch` like:

https://github.com/WordPress/gutenberg/pull/4839#issuecomment-364673030

**Comment by jasmussen** [comment]
This seems like an impressive enhancement, though it's a little difficult for me to understand exactly what goes on. Perhaps you can elaborate a bit? 

What seems to happen is that even if I try to embed an invalid URL, it creates an embed of sorts. It would seem more obvious to me if we showed an error here, perhaps. 

I noticed an issue, btw, not sure if this is both on master and in this branch:

autosave

Steps to reproduce:

1. Insert video block in fresh post
2. Paste video URL, valid or invalid, doesn't matter
3. Wait for autosave, or invoke it manually
4. Reload the page
5. Insert video block again

The second time you insert a video block, it's born with the URL you pasted previously.

**Comment by mcsf** [comment]
Thanks for having a look at this, @jasmussen, despite my inability to get back to you. :)

The parent issue has been closed, so I'm dropping this one.

**Pull request closed without merging**